### PR TITLE
chore: update react-instancesearch to v7

### LINKED ISSUE #55410: Upgrade React InstantSearch to v7
## Description

We are using React InstantSearch v6 for the search bar on /learn, and this version is [deprecated](https://www.algolia.com/doc/deprecated/instantsearch/react/v6/api-reference/instantsearch).

Migration guide: https://www.algolia.com/doc/guides/building-search-ui/upgrade-guides/react/#migrate-from-react-instantsearch-v6-to-react-instantsearch-v7.

### PR DESCRIPTION
**Pull request by Sembauke**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #55410

### CODE REVIEW

**Review by huyenltnguyen — COMMENTED** [review]
I finally had the courage to click on this PR 😛 

Not sure what made you feel like picking this one, but that means less hair-pulling for me, so thank you 😄 

---
I added some small changes to the PR. I'll need to go through the PR again to digest the changes, though I tested the component locally and things seem to be working as intended.

File: client/src/components/search/searchBar/search-bar.tsx [hunk 1]
```diff
@@ -115,20 +114,19 @@ export class SearchBar extends Component<SearchBarProps, SearchBarState> {
...
     toggleSearchDropdown(false);
     if (selectedHit) {
       // Redirect to hit / footer selected by arrow keys
-      return window.location.assign(selectedHit.url);
+      return window.location.assign(selectedHit.url as string);
     } else if (!query) {
       // Set query to value in search bar if enter is pressed
       query = (e.currentTarget?.children?.[0] as HTMLInputElement).value;
     }
-    updateSearchQuery(query);
```
[the remark this answers is not in the document]
Comment (Sembauke) [reply client/src/components/search/searchBar/search-bar.tsx:131]: Yes, this was fully intentional!

**Review by huyenltnguyen — COMMENTED** [review]
I just have a couple of questions left. But everything else looks good to me 👍

**Review by huyenltnguyen — APPROVED** [review]

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Other than the highlighting, this seems to work well. I don't think we need to force the types, though:

**Comment by Sembauke** [comment]
@ojeytonwilliams I agree, could you open an issue for that? 

All changes have been applied.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:, thanks @Sembauke

**Pull request merged**